src/hotspot/os/windows/os_windows.cpp

```diff
@@ -867,21 +868,60 @@ int os::active_processor_count() {
     return ActiveProcessorCount;
   }
 
+  bool schedules_all_processor_groups = false;
+  if (UseAllWindowsProcessorGroups) {
+    // Starting with Windows 11 and Windows Server 2022, the OS has changed to
+    // make processes and their threads span all processors in the system,
+    // across all processor groups, by default. Therefore, we will allow all
+    // processors to be active processors on these operating systems. However,
+    // job objects can be used to restrict processor affinity across the
+    // processor groups. In this case, the number of active processors must be
+    // obtained from the processor affinity in the job object.
+    schedules_all_processor_groups = win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater();
+    if (schedules_all_processor_groups) {
+      DWORD processors_in_job_object = win32::active_processors_in_job_object();
+
+      if (processors_in_job_object > 0) {
+        return processors_in_job_object;
+      }
+    }
+  }
+
+  DWORD logical_processors = 0;
+  SYSTEM_INFO si;
+  GetSystemInfo(&si);
+
+  // There is no associated job object so get the number of available logical processors from the process affinity mask
   DWORD_PTR lpProcessAffinityMask = 0;
   DWORD_PTR lpSystemAffinityMask = 0;
-  int proc_count = processor_count();
-  if (proc_count <= sizeof(UINT_PTR) * BitsPerByte &&
-      GetProcessAffinityMask(GetCurrentProcess(), &lpProcessAffinityMask, &lpSystemAffinityMask)) {
-    // Nof active processors is number of bits in process affinity mask
-    int bitcount = 0;
-    while (lpProcessAffinityMask != 0) {
-      lpProcessAffinityMask = lpProcessAffinityMask & (lpProcessAffinityMask-1);
-      bitcount++;
-    }
-    return bitcount;
+  if (GetProcessAffinityMask(GetCurrentProcess(), &lpProcessAffinityMask, &lpSystemAffinityMask)) {
```

Comment:
> I'm struggling to see how we can expose more than 64 processors when our bit-vector is only 64 bits. ??

This code block does not expose over 64 processors. It only ensures that a request for less than 64 processors is respected when the user launches the process using the start command (which accepts a 64-bit mask) https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/start

More than 65 cores are exposed by the `logical_processors = processor_count();` assignment a few lines later.

Reply:
Sorry I'm still not clear here. From my reading `GetProcessAffinityMask` is either going to return a zero mask or a non-zero mask. In the non-zero case, the count must be <=64. So even if we intend to use > 64 processors, this mask reading will always limit us to a value < 64. And the zero mask is only for one special case. So regardless of the use of the "start" command, it seems to me this will cause us to bail out at line 903. ??

Reply:
> Sorry I'm still not clear here. From my reading `GetProcessAffinityMask` is either going to return a zero mask or a non-zero mask. In the non-zero case, the count must be <=64. So even if we intend to use > 64 processors, this mask reading will always limit us to a value < 64. And the zero mask is only for one special case. So regardless of the use of the "start" command, it seems to me this will cause us to bail out at line 903. ??

If there are 64 bits set in the process affinity mask, there is no way to tell whether or not the start command was used to set this mask. This implementation treats this case as equivalent to the user not restricting processor affinities at all. We don't bail out at line 903 because in this case, logical_processors == si.dwNumberOfProcessors and all 65+ processors from line 964 will therefore be usable.

Reply:
Sorry to belabour this but IIUC the only time we will use > 64 processors is if the affinity masks is all ones - right? But we cannot distinguish between the cases of only wanting to use 64 processors and wanting to use > 64 processors. And if we didn't use the start command and have not otherwise set the affinity mask, then that mask will be all ones.